src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -910,6 +933,90 @@ private void CreateFirstSegment(string segmentText)
+        /// <param name="segmentText">The text of the segment.</param>
+        private bool CreateSegmentForEscapeFunction(string segmentText)
+        {
+            string identifier;
+            string parenthesisExpression;
+
+            ExtractSegmentIdentifierAndParenthesisExpression(segmentText, out identifier, out parenthesisExpression);
+
+            if (this.parsedSegments.Count == 0)
+            {
+                if (this.TryCreateSegmentForNavigationSource(identifier, parenthesisExpression))
+                {
+                    return true;
+                }
+
+                if (this.TryCreateSegmentForOperationImport(identifier, parenthesisExpression))
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                if (this.TryCreateFilterSegment(segmentText))
+                {
+                    return true;
+                }
+
+                ODataPathSegment previous = this.parsedSegments[this.parsedSegments.Count - 1];
+
+                // property if previous is single
+                if (previous.SingleResult)
+                {
+                    // if its not one of the recognized special segments, then it must be a property, type-segment, or key value.
+                    Debug.Assert(
+                        previous.TargetKind == RequestTargetKind.Resource
+                        || previous.TargetKind == RequestTargetKind.Dynamic,
+                        "previous.TargetKind(" + previous.TargetKind + ") can have properties");
+
+                    if (previous.TargetEdmType == null)
+                    {
+                        // A segment will correspond to a property in the object model;
+                        // if we are processing an open type, anything further in the
+                        // URI also represents an open type property.
+                        Debug.Assert(previous.TargetKind == RequestTargetKind.Dynamic, "For open properties, the target resource type must be null");
+                    }
+                    else
+                    {
+                        // if the segment corresponds to a declared property, handle it
+                        // otherwise, fall back to type-segments, actions, and dynamic/open properties
+                        IEdmProperty projectedProperty;
+                        if (this.TryBindProperty(identifier, out projectedProperty))
+                        {
+                            CheckSingleResult(previous.SingleResult, previous.Identifier);
+                            this.CreatePropertySegment(previous, projectedProperty, parenthesisExpression);
+
+                            return true;
+                        }
+                    }
+                }
+
+                // Type cast
+                if (segmentText.IndexOf('.') >= 0 && // type-cast should use qualified type names

```

Comment:
>segmentText.IndexOf('.') >= 0 [](start = 20, length = 29)

should we move this check into TryCreateTypeNameSegment?

Reply:
I would have agreed with you. I had two thoughts stopping me - 
1. Either I change the parameters of the function to take in the segment text or do the same check on identifier. I am worried as there might be a case where two are not the same.
2. Adding another function to the call stack on the hot path will add to COGS. I have a better idea for performance, which is to check for this in extractsegmentidentifier method, I will add that in other PR. 

---
In reply to: [369793950](https://github.com/OData/odata.net/pull/1621#discussion_r369793950) [](ancestors = 369793950)

Reply:
Interesting... only case where segmentText would be different from identifier would be if it included parens, right?  In that case, shouldn't we be be checking identifier, rather than segmentText anyway?

---
In reply to: [369824859](https://github.com/OData/odata.net/pull/1621#discussion_r369824859) [](ancestors = 369824859,369793950)